### PR TITLE
Provide treeView service API

### PR DESCRIPTION
**Pull request by UziTech**
### Description of the Change

Provided a service that lets other packages consume the treeView instance.

At the moment any package that interacts with the tree-view package to get selected files or other information are doing so by getting the active package's `mainModule` or checking classes and data attributes on elements.

There is a standard way to get this information through the [Services API](http://flight-manual.atom.io/behind-atom/sections/interacting-with-other-packages-via-services/)

### Alternate Designs

We may want to change what actually get provided if there are unstable features in the tree-view module.

### Benefits

Other packages can interact with this package in a standard way.

### Possible Drawbacks

none

### Applicable Issues

#433

### CODE REVIEW

**Comment by UziTech** [comment]
Here is an example of why this would be useful:

https://github.com/subesokun/atom-tree-view-git-status/blob/master/lib/main.coffee#L154

The `tree-view-git-status` package uses 

```javascript
atom.packages.getActivePackage('tree-view').mainModule
```

to get the treeView instance. This poses two problems:

  1. It will break if the module changes the way it stores the treeView instance. (Ass it seems to have done in v1.18.0)
  2. It limits `tree-view-git-status` to only work with the `tree-view` package instead of also working with any fork of the `tree-view` package.

**Comment by thomasjo** [comment]
Thanks for the contribution! I do however have concerns. Your proposition exposes the entire tree-view instance, which does not feel like an appropriate solution. The only justification I can think of is that it's the easiest approach with the least amount of effort. Keep in mind that most (if not all) of the tree-view object consists of private methods.

I propose that we expose an actual API, instead of simply handing developers an entire object, which essentially has no lifetime guarantees etc. For instance, if someone relies on this object, and the tree-view package is disabled, what do you think/expect will happen?

To begin with, which aspects of the tree-view do you think we have to expose? Let's begin with the bare minimum of features that are actually needed by external packages rights now — YAGNI™ and so on.

**Comment by UziTech** [comment]
I agree, exposing the entire tree-view instance is probably not the right way to start out.

There are already some methods in tree-view class that are marked [public](https://github.com/atom/tree-view/blob/master/lib/tree-view.coffee#L843). I'm assuming those are supposed to be public to the rest of the package. Those seem unlikely to change in any major way.

The most common thing other packages seem to do is get the selected paths. If we wanted to start really small we should start with that.

**Comment by thomasjo** [comment]
> The most common thing other packages seem to do is get the selected paths. If we wanted to start really small we should start with that.

OK, that sounds like a reasonable approach :1st_place_medal: 
Expose that functionality via a wrapper object. Ideally something that can withstand the tree-view being disabled, but doing something simple like we do in status-bar is probably acceptable as well. We can always improve that part later without breaking any dependent package.

**Comment by UziTech** [comment]
I added `getEntry` as well for packages like `tree-view-git-status` that want to edit the elements in tree view.

`getEntry` sounds kind of weird to me. Should I change it to `getElement` or something like that?

File: lib/tree-view-package.js [hunk 1]
```diff
@@ -52,6 +52,13 @@ class TreeViewPackage {
     )
   }
 
+  provideTreeView () {
+    return {
+      getSelectedPaths: () => this.getTreeViewInstance().selectedPaths(),
+      getEntry: (path) => this.getTreeViewInstance().entryForPath(path)
```
Comment (thomasjo) [inline lib/tree-view-package.js]: I am one of those people who dislike the `get` and `set` prefixes, since both words are extremely difficult to summarize/describe — look up both words in any English dictionary, and you will see what I mean (see e.g. [get](http://dictionary.cambridge.org/dictionary/english/get), and [set](http://dictionary.cambridge.org/dictionary/english/set)).

**TL; DR** What about changing these `selectedPaths` and `entryForPath`, just as they are named in the underlying API.

Comment (UziTech) [reply lib/tree-view-package.js]: I agree using the same name as the treeView instance might be helpful

**Comment by UziTech** [comment]
@thomasjo anything else holding this up?

File: spec/tree-view-package-spec.coffee [hunk 2]
```diff
@@ -3847,6 +3847,23 @@ describe "TreeView", ->
     files = Array.from(element.querySelectorAll('.entries .file'))
     files.find((file) -> file.fileName.textContent is text)
 
+describe "Service provider", ->
+  [provideTreeView, treeView] = []
+  beforeEach ->
+    waitForPackageActivation()
+
+    runs ->
+      provideTreeView = atom.packages.getActivePackage('tree-view').mainModule.provideTreeView()
```
Comment (thomasjo) [inline spec/tree-view-package-spec.coffee]: Change the name of this to `treeViewService`, and move it below `treeView` (also re-order the variable declarations above).

File: spec/tree-view-package-spec.coffee [hunk 3]
```diff
@@ -3847,6 +3847,23 @@ describe "TreeView", ->
...
+      treeView = atom.workspace.getLeftDock().getActivePaneItem()
+
+  it "should provide selected paths", ->
+    expect(provideTreeView.selectedPaths()).toEqual([atom.project.getPaths()[0]])
+
+  it "should provide entry for a path", ->
```
Comment (thomasjo) [inline spec/tree-view-package-spec.coffee]: The description of both of these tests should be a bit more meaningful I think, but sadly I am unable to churn out anything right now to help you. Maybe one of the other maintainers can chime in with some ideas on how to rewrite them.

/cc @atom/feedback

File: package.json [hunk 4]
```diff
@@ -30,6 +30,14 @@
       }
     }
   },
+  "providedServices": {
+    "tree-view": {
+      "description": "Get treeView Instance",
```
Comment (thomasjo) [inline package.json]: Rewrite this to something more descriptive and helpful to developers; perhaps something like:
"A tree-like view of directories and files"

/cc @atom/feedback

File: spec/tree-view-package-spec.coffee (hunk 3 above)
Comment (UziTech) [reply spec/tree-view-package-spec.coffee]: I'm not sure what this should test other than that the method exists on the service. The internal methods should have their own tests to make sure they are actually working properly.

**Comment by thomasjo** [comment]
This LGTM! Thanks for not giving up, @UziTech :stuck_out_tongue_winking_eye: 

---
@atom/maintainers I'll give everyone a few hours to chime in on this, then I'll merge and ship this.

**Comment by nathansobo** [comment]
@thomasjo You cool documenting this in the README before merging?

**Comment by thomasjo** [comment]
@nathansobo Sure thing! Good call 🥇

**Comment by thomasjo** [comment]
@UziTech You mind providing Atom maintainers push access to your fork? That wait I can push my commits directly to your fork before merging.

**Comment by UziTech** [comment]
@thomasjo "Allow edits from maintainers." is checked is there anything else I need to do?

**Comment by thomasjo** [comment]
Travis is being slow, so I'm going to merge this since the build was green prior to my documentation (and argument name tweak).

@UziTech Thanks for this! 🙇

**Comment by Alhadis** [comment]
Well now, looks like we're another step closer to `tree-view` services providing DOM access. ;) Perhaps [this](https://github.com/atom/tree-view/pull/1146) should be the next API addition to review?
